### ISSUE TITLE
Add IArgument to AMP

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3468]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3468|
|Type|enhancement|
|Created|2008-10-13 14:04:43Z|
|Branch|https://github.com/twisted/twisted/tree/iargument-3468|

AMP doesn't define very many types of arguments.  It's expected and good for applications to define their own argument types.

However, it's not clear what the API for defining a new argument is.  It seems as though there should be an `IArgument` which defines this.

<details><summary>Searchable metadata</summary>

```
trac-id__3468 3468
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_iargument_3468 branches/iargument-3468
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1223906683000000 1223906683000000
changetime__1231028618000000 1231028618000000
version__None None
owner__ 
cc__glyph
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [25754]) Branching to 'iargument-3468'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @exarkun|
|-|-|

Looks pretty nice so far, but I have a few suggestions:

  * `IArgument` sounds nicer than `IArgumentType`.  `Type` on the end of an interface name is usually a code smell.  A provider of this interface will represent a specific argument of a specific command; one is expected to instantiate it.  In other words, the implementor is the argument type, the instance is the argument.
  * replace `"L{IArgumentType} providers define a way of representing"` with `"An L{IArgument} can serialize"`.  I like to avoid words like "provider", "represent" in interface docstrings.
  * A better test would verify `String`, `Unicode`, `Integer`, et. al.; `Argument` doesn't _actually_ implement this interface; see, for example, the disparity between the implementations of `Argument.fromString` and `Argument.toString` and their docstrings.

I'd say that even as-is it would be an improvement though, so address as much of this as you like and then go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

  1. The object which provides this interface isn't an "argument".  To be an argument, I think it needs to have a name, but it doesn't.  The only thing the object which provides this interface does is implement serialization and deserialization for an argument.  That makes it feel like an argument type to me.  It's like the `int` int `void foo(int x);`.  I agree that `IArgumentType` may not be the ideal name, but I don't think `IArgument` is better.  Maybe there's a third possibility?
  1. Changed in r25766
  1. I would say _A better set of tests_ rather than _A better test_, because I would leave the test I added and write more for `String`, `Unicode`, etc, as these seem to be somewhat lacking in good unit test coverage.  But I don't think I changed these classes significantly in this branch so I don't feel compelled to add more tests before merging this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_4):
>   1. The object which provides this interface isn't an "argument".  To be an argument, I think it needs to have a name, but it doesn't.  The only thing the object which provides this interface does is implement serialization and deserialization for an argument.  That makes it feel like an argument type to me.  It's like the `int` int `void foo(int x);`.  I agree that `IArgumentType` may not be the ideal name, but I don't think `IArgument` is better.  Maybe there's a third possibility?

It's in the right spot to have a name.  I'd very much like to expand the interface to give it a name, in the future.  I think `[('x', Y())]` was a mistake, and it really should have been `[Y('x')]`.  It's syntactically more convenient and it keeps information together in a more useful form.  If you like, we can make a separate ticket for this change and have a new `IArgument` interface later, when those changes have been made.

>   2. Changed in r25766

Great.

>   3. I would say _A better set of tests_ rather than _A better test_, because I would leave the test I added and write more for `String`, `Unicode`, etc, as these seem to be somewhat lacking in good unit test coverage.  But I don't think I changed these classes significantly in this branch so I don't feel compelled to add more tests before merging this.

Indeed.  This was really just nitpicking; if we are going to use this interface for anything later, we can have tests for whatever we're using it for.  You are welcome to ignore this feedback.

So, address point 1 to your satisfaction and go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

I like the idea of introducing a new interface later to collapse the argument name/type into a single object.  So, let's do that, and I'll leave `IArgumentType` as it is.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [25780]) Merge iargument-3468 

Author: exarkun
Reviewer: glyph
Fixes: [#3468](https://github.com/twisted/twisted/issues/3468)

Add `IArgumentType`, an interface which defines the requirements for the
second element of the two-tuples in the `arguments` and `response` lists
of `Command` and its subclasses.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
